### ISSUE TITLE
Json escape

### ISSUE THREAD

**Issue by lehoanglu0ng** [apollographql/apollo-android#1848]
I try to make a request with JSON variable like this (Strapi syntax):
```
query Test ($condition: JSON!) {
    foods(where: $condition) {
        name
     }
}
```
But the request payload is parsed to 
```
{"operationName":"Test","variables":{"condition":"{\"name_contains\":\"abc\"}"},"query":"query Test($condition: JSON!)"}}
```
which should be
```
{"operationName":"Test","variables":{"condition":"{"name_contains":"abc"}"},"query":"query Test($condition: JSON!)"}}
```
After investigating, I notice that the root cause is apollo parsed the JSON variable to `Map<String, Object>`
**Version**
1.2.2

**Comment by cfsnsalazar** [comment]
I am having the same issue

**Comment by martinbonnin** [comment]
IIRC, `JSON` is not part of the spec, right ? So it's either a Custom Scalar Type or a type defined in your schema ? 

1. If it's a Custom Scalar Type, can you share your adapter ?
2. If it's defined in your schema, can you share the definition from schema.json ?

**Comment by cfsnsalazar** [comment]
It's a custom scalar type. I've been using Jackson to deal with raw JSON from the beginning. It seems to be encoded the json fine at the adapter level.

`val jsonTypeAdapter = object : CustomTypeAdapter<JsonNode> {
  override fun encode(value: JsonNode): CustomTypeValue<*> {
    return when (value.nodeType) {
      ARRAY -> {
        val arrayNode = value as? ArrayNode
        CustomTypeValue.GraphQLJsonList(value.elements().asSequence<Any>().toMutableList())
      }
      STRING -> CustomTypeValue.GraphQLString(value.toString())
      OBJECT,
      POJO -> CustomTypeValue.GraphQLJsonObject(JSONMapperUtil.createMapByObject(value) as Map<String, Any>)
      else -> CustomTypeValue.fromRawValue("")
    }
  }

  override fun decode(value: CustomTypeValue<*>): JsonNode {
    return JSONMapperUtil.createJSONNodeByObject(value.value)
  }
}`

Also, these are my custom type definitions
`customTypeMapping = [
      "DateTime": "org.joda.time.DateTime",
      "JSON": "com.fasterxml.jackson.databind.JsonNode",
      "GeoJSON": "org.geojson.GeoJsonObject",
      "UniqueID": "kotlin.String",
      "Upload": "com.apollographql.apollo.api.FileUpload",
      "Regex": "kotlin.String"
  ]`

<img width="184" alt="Screen Shot 2019-12-20 at 9 37 53 AM" src="https://user-images.githubusercontent.com/16261373/71265797-9f744b80-230c-11ea-9f5f-f347b8d31a41.png">

I would attach my schema.json file, but it's probably too big to be attached here

**Comment by sav007** [comment]
Question: in you adapter implementation I see that you encode `JsonNode` into `CustomTypeValue.GraphQLJsonObject`. Why not encode it as string (stringify json) `CustomTypeValue.GraphQLString` ?

**Comment by m-Skolnick** [comment]
@sav007 I am working on the same project as @cfsnsalazar. I tried encoding the JsonNode as a string, but the quotation marks are still being escaped whether it is `CustomTypeValue.GraphQLJsonObject` or `CustomTypeValue.GraphQLString`.

**Comment by m-Skolnick** [comment]
@sav007 After a little bit of digging, I believe the issue is in `JsonUtf8Writer` found at `com.apollographql.apollo.api.internal.json`. Quotation marks are in the list of unicode characters which must be escaped. Also this comment can be found at the top of the file. Other than that, I do not see any other explanation for why quotation marks must be escaped. 
```  /*
   * From RFC 7159, "All Unicode characters may be placed within the
   * quotation marks except for the characters that must be escaped:
   * quotation mark, reverse solidus, and the control characters
   * (U+0000 through U+001F)."
   *
   * We also escape '\u2028' and '\u2029', which JavaScript interprets as
   * newline characters. This prevents eval() from failing with a syntax
   * error. http://code.google.com/p/google-gson/issues/detail?id=341
   */```

**Comment by sav007** [comment]
I tried to reproduce this issue with no luck.

@m-Skolnick if you take a look at how custom scalar types are serialized before sending, there is a branch:
https://github.com/apollographql/apollo-android/blob/master/apollo-api/src/main/java/com/apollographql/apollo/api/internal/json/InputFieldJsonWriter.java#L94-L97

and here https://github.com/apollographql/apollo-android/blob/master/apollo-api/src/main/java/com/apollographql/apollo/api/internal/json/Utils.java#L13-L21

you can see that if your encoder returns `CustomTypeValue.GraphQLJsonObject` then it will be serialized as json object rather than `stringified` json string.

I was trying to reproduce the issue with following:
gradle config:
```
customTypeMapping.set(mapOf(
      "JSON" to "java.util.Map"
  ))
```
Apollo supports mapping to `java.util.Map` out of the box, so no need to provide custom scalar type adapter https://github.com/apollographql/apollo-android/blob/master/apollo-api/src/main/java/com/apollographql/apollo/response/ScalarTypeAdapters.java#L122-L130

graphql query:
```
query Hero($condition: JSON!) {
    hero(where: $condition) {
        name
    }
}
```

test:
```
val query = HeroQuery(mapOf("name" to mapOf<String, Any>("equal" to "abc")))
```

request payload sent to server was as expected:
```
{
  "operationName":"Hero",
  "variables":{"condition":{"name":{"equal":"abc"}}},
  "query":"query Hero($condition: JSON!) { hero(where: $condition) { __typename name } }"
}
```

One suggestion is could please try to debug `InputFieldJsonWriter` and see how your custom scalar type is serialized.

**Comment by cfsnsalazar** [comment]
Hello @sav007, sorry for being absent for the past couple of weeks. I was on Christmas break. Btw, happy 2020!.

So, we could change our models to use `Map` when trying to parse JSON. However, I believe this is not a final solution since it might not cover all the scenarios that we need to cover. 

Having that being said, I have a question in regards to `InputFieldJsonWriter`:

why do you pass `value` to `Utils.writeToJson(value, jsonWriter)` instead of `customTypeValue.value`? .

I noticed that `Utils.writeJson` defaults to `jsonWriter.value(value.toString())` if the value being passed is not a `Map`,  a `List`, a `Boolean`, or a `Number`. A Map could be passed to the utils by using `customTypeValue.value`. However, I am not sure if there is a reason to pass `value` instead

**Comment by sav007** [comment]
>why do you pass value to Utils.writeToJson(value, jsonWriter) instead of customTypeValue.value? 

Well it's I guess it's because `JsonWriter`  doesn't support writing for Map and Lists. To write them to JSON we need to do this:
```
jsonWriter.beginObject();
// write map that can have nested objects (maps) or lists
jsonWriter.endObject();
```

Agh, now I guess I have an idea of the issue. If you pass a map that contains values that is not supported by JSON then it will be converted to string as a fallback. Interesting...

Not sure though how to handle that properly, but the idea that passed map MUST contain only these types: `Map, List, Boolean, Number, String`.

**Comment by sav007** [comment]
Seems like this is not an issue but rather we should have better documentation for GraphQL custom scalar type support

**Comment by cfsnsalazar** [comment]
After 1.2.3 , I can tell that our issue was fixed with commit 6ecc218
